packages/design-system/src/components/ButtonIcon/ButtonIconVariant.tsx

```diff
@@ -0,0 +1,40 @@
+import React from 'react';
+import ButtonIcon from './variations/ButtonIcon';
+import ButtonToggle from './variations/ButtonToggle';
+import ButtonFloating from './variations/ButtonFloating';
+import { ButtonIconProps } from './Primitive/ButtonIconPrimitive';
+
+const ButtonIconVariant = React.forwardRef(
+	(props: ButtonIconProps, ref: React.Ref<HTMLButtonElement>) => {
+		switch (props.variant) {
+			case 'toggle': {
+				const { children, variant, ...rest } = props;
```

Comment:
How about having this line before the switch to have it only one time?

Reply:
When you decompose the object TS can't infer type anymore, so inside each case it wouldn't be able to know that the props are correctly typed for the given variant. 

It made me sad as well.